File: tests/oracle/stop_ptrsub_wire_1204.cc
/* STOP-PTRSUB-WIRE-0001: ActionInferTypes STOP seal + PTRSUB downChain wiring.
 *
 * Bilateral fixture: one Funcdata drives ActionInferTypes through the
 * production pipeline on a graph that exercises
 *   - buildLocaltypes' needsBlock -> setStopUpPropagation
 *     (coreaction.cc:5030-5031) for a PTRSUB flagged stop_type_propagation,
 *   - propagateTypeEdge's stopsUpPropagation guard (coreaction.cc:5093),
 *   - TypeOpPtrsub::propagateType -> propagateAddIn2Out -> downChain
 *     (typeop.cc:2375/1215) producing the field pointer + ephemeral
 *     getTypePointerRel form (type.cc:4016),
 *   - the INT_ADD pointer arm (typeop.cc:1200) and the INT_SUB null arm
 *     (typeop.cc:317-321).
 */
#include <bits/stdc++.h>

// Test-only access rewrite for friend-only setters. Standard headers are
// included first so the visibility rewrite never touches libstdc++.
#define private public
#include "architecture.hh"
#include "coreaction.hh"
#include "funcdata.hh"
#include "libdecomp.hh"
#include "marshal.hh"
#include "op.hh"
#include "translate.hh"
#include "type.hh"
#include "typeop.hh"
#include "varnode.hh"
#undef private

using namespace std;
using namespace ghidra;

class FixtureTranslate final : public Translate {
  VarnodeData dummyRegister;

public:
  FixtureTranslate()
  {
    setBigEndian(false);
    setUniqueBase(0);
    insertSpace(new ConstantSpace(this, this));
    insertSpace(new AddrSpace(this, this, IPTR_PROCESSOR, "other", false,
                              8, 1, 1, AddrSpace::hasphysical, 0, 0));
    insertSpace(new UniqueSpace(this, this, 2, 0));
    AddrSpace *ram = new AddrSpace(this, this, IPTR_PROCESSOR, "ram", false,
                                   8, 1, 3, AddrSpace::hasphysical, 0, 0);
    insertSpace(ram);
    AddrSpace *reg = new AddrSpace(this, this, IPTR_PROCESSOR, "register", false,
                                   8, 1, 4, AddrSpace::hasphysical, 0, 0);
    insertSpace(reg);
    SpacebaseSpace *stack = new SpacebaseSpace(
        this, this, "stack", 5, 8, ram, 1, true);
    insertSpace(stack);
    VarnodeData stackPointer;
    stackPointer.space = reg;
    stackPointer.offset = 0;
    stackPointer.size = 8;
    addSpacebasePointer(stack, stackPointer, 8, true);
    insertSpace(new AddrSpace(this, this, IPTR_PROCESSOR, "join", false,
                              8, 1, 6, AddrSpace::hasphysical, 0, 0));
    insertSpace(new IopSpace(this, this, 7));
    setDefaultCodeSpace(3);
    dummyRegister.space = reg;
    dummyRegister.offset = 0;
    dummyRegister.size = 8;
  }

  void initialize(DocumentStorage &) override {}
  const VarnodeData &getRegister(const string &) const override {
    return dummyRegister;
  }
  string getRegisterName(AddrSpace *, uintb, int4) const override { return ""; }
  string getExactRegisterName(AddrSpace *, uintb, int4) const override { return ""; }
  void getAllRegisters(map<VarnodeData, string> &) const override {}
  void getUserOpNames(vector<string> &) const override {}
  int4 instructionLength(const Address &) const override { return 0; }
  int4 oneInstruction(PcodeEmit &, const Address &) const override { return 0; }
  int4 printAssembly(AssemblyEmit &, const Address &) const override { return 0; }
};

class FixtureArchitecture final : public Architecture {
protected:
  Translate *buildTranslator(DocumentStorage &) override { return (Translate *)0; }
  void buildLoader(DocumentStorage &) override {}
  PcodeInjectLibrary *buildPcodeInjectLibrary(void) override {
    return (PcodeInjectLibrary *)0;
  }
  void buildTypegrp(DocumentStorage &) override {}
  void buildCoreTypes(DocumentStorage &) override {}
  void buildCommentDB(DocumentStorage &) override {}
  void buildStringManager(DocumentStorage &) override {}
  void buildConstantPool(DocumentStorage &) override {}
  void buildContext(DocumentStorage &) override {}
  void buildSymbols(DocumentStorage &) override {}
  void buildSpecFile(DocumentStorage &) override {}
  void modifySpaces(Translate *) override {}
  void resolveArchitecture(void) override {}

public:
  FixtureArchitecture()
  {
    FixtureTranslate *fixtureTranslate = new FixtureTranslate();
    translate = fixtureTranslate;
    copySpaces(fixtureTranslate);
    max_basetype_size = 10;
    types = new TypeFactory(this);
    types->setupSizes();
    types->setCoreType("xunknown1", 1, TYPE_UNKNOWN, false);
    types->setCoreType("xunknown2", 2, TYPE_UNKNOWN, false);
    types->setCoreType("xunknown4", 4, TYPE_UNKNOWN, false);
    types->setCoreType("xunknown8", 8, TYPE_UNKNOWN, false);
    types->cacheCoreTypes();
    TypeOp::registerInstructions(inst, types, translate);
    symboltab = new Database(this, false);
    symboltab->attachScope(new ScopeInternal(0x101, "", this), (Scope *)0);
    ProtoModel *model = new ProtoModel(this);
    istringstream stream(
        "<prototype name=\"fixture\" extrapop=\"0\"><input/><output/></prototype>");
    XmlDecode decoder(this);
    decoder.ingestStream(stream);
    model->decode(decoder);
    protoModels[model->getName()] = model;
    setDefaultModel(model);
  }

  void printMessage(const string &) const override {}
};

class FixtureProgress final : public TypeStruct {
public:
  explicit FixtureProgress(const vector<TypeField> &fields)
  {
    name = "ProgressData";
    displayName = name;
    setFields(fields, 32, 8);
    markComplete();
  }
};

static string metaToken(type_metatype meta)
{
  switch (meta) {
  case TYPE_VOID: return "void";
  case TYPE_PTR: return "ptr";
  case TYPE_ARRAY: return "array";
  case TYPE_STRUCT: return "struct";
  case TYPE_SPACEBASE: return "spacebase";
  case TYPE_BOOL: return "bool";
  case TYPE_INT: return "int";
  case TYPE_UINT: return "uint";
  default: return "other";
  }
}

/// Normalized projection shared with the Rust side: base types print as
/// metatype+size tokens; pointers print the pointed-to projection plus the
/// relative-pointer container state (parent name and byte offset) when the
/// dynamic type is TypePointerRel.
static string typeProj(const Datatype *ct)
{
  if (ct == (const Datatype *)0) return "null";
  if (ct->getMetatype() == TYPE_PTR) {
    const TypePointer *p = (const TypePointer *)ct;
    string res = "ptr" + to_string(p->getSize()) + "->" + typeProj(p->getPtrTo());
    const TypePointerRel *rel = dynamic_cast<const TypePointerRel *>(ct);
    if (rel != (const TypePointerRel *)0)
      res += "|rel=" + rel->getParent()->getName() + ":" + to_string(rel->getByteOffset());
    return res;
  }
  return metaToken(ct->getMetatype()) + to_string(ct->getSize());
}

struct Cell {
  string label;
  Varnode *vn;
};

static string snapshot(const vector<Cell> &cells)
{
  ostringstream stream;
  for (size_t i = 0; i < cells.size(); ++i) {
    if (i != 0) stream << ',';
    stream << cells[i].label << ':'
           << typeProj(cells[i].vn->getType())
           << "/stop=" << (cells[i].vn->stopsUpPropagation() ? 1 : 0);
  }
  return stream.str();
}

int main(void)
{
  cout << std::unitbuf;
  AttributeId::initialize();
  ElementId::initialize();
  CapabilityPoint::initializeAll();
  ArchitectureCapability::sortCapabilities();
  cout << "schema=1|fixture=STOP-PTRSUB-WIRE-0001"
       << "|oracle=e40ed13014025f82488b1f8f7bca566894ac376b\n";
  try {
    FixtureArchitecture architecture;
    AddrSpace *ram = architecture.getSpace(3);
    AddrSpace *reg = architecture.getSpace(4);
    Scope *global = architecture.symboltab->getGlobalScope();
    Funcdata fd("stop_ptrsub_wire", "stop_ptrsub_wire", global,
                Address(ram, 0x5000), (FunctionSymbol *)0, 0x40);
    BlockGraph &blocks = const_cast<BlockGraph &>(fd.getBasicBlocks());
    BlockBasic *block = blocks.newBlockBasic(&fd);
    fd.setBasicBlockRange(block, Address(ram, 0x5000), Address(ram, 0x5009));

    Datatype *longType = architecture.types->getBase(8, TYPE_INT);
    vector<TypeField> fields;
    fields.push_back(TypeField(0, 0, "total", longType));
    fields.push_back(TypeField(1, 8, "prev", longType));
    fields.push_back(TypeField(2, 16, "point", longType));
    TypeStruct *progress = new FixtureProgress(fields);
    Datatype *progressPointer =
        architecture.types->getTypePointer(8, progress, 1);

    Varnode *bar = fd.newVarnode(8, Address(reg, 0x100));
    bar = fd.setInputVarnode(bar);
    bar->updateType(progressPointer, true, false);

    // LOAD space constants carry the AddrSpace POINTER as the constant
    // offset: VarnodeData::getSpaceFromConst (pcoderaw.hh:101-103) casts the
    // offset straight back to `AddrSpace*`, and TypeOpLoad::propagateType's
    // inslot==-1 arm dereferences it.
    Varnode *spaceConst = fd.newConstant(4, (uintb)(uintp)ram);
    vector<Cell> cells;
    int4 seq = 0;

    // Helper lambdas building each case's op chain.
    auto makePtrsub = [&](uintb off, bool stop) -> Varnode * {
      PcodeOp *op = fd.newOp(2, Address(ram, 0x5000 + (seq++)));
      fd.opSetOpcode(op, CPUI_PTRSUB);
      fd.opSetInput(op, bar, 0);
      fd.opSetInput(op, fd.newConstant(8, off), 1);
      Varnode *out = fd.newUniqueOut(8, op);
      fd.opInsertEnd(op, block);
      if (stop)
        op->setStopTypePropagation();
      return out;
    };
    auto makeLoad = [&](Varnode *addr) -> Varnode * {
      PcodeOp *op = fd.newOp(2, Address(ram, 0x5000 + (seq++)));
      fd.opSetOpcode(op, CPUI_LOAD);
      fd.opSetInput(op, spaceConst, 0);
      fd.opSetInput(op, addr, 1);
      Varnode *out = fd.newUniqueOut(8, op);
      fd.opInsertEnd(op, block);
      return out;
    };

    // A: sealed PTRSUB (stop flag on the op, as RulePtrArith sets it) + LOAD.
    Varnode *tA = makePtrsub(8, true);
    Varnode *vA = makeLoad(tA);
    // B: identical PTRSUB without the stop flag (control).
    Varnode *tB = makePtrsub(8, false);
    Varnode *vB = makeLoad(tB);
    // C: INT_SUB consumes the pointer — base TypeOp::propagateType returns
    // null (typeop.cc:317-321): no pointer reaches the output.
    PcodeOp *subOp = fd.newOp(2, Address(ram, 0x5000 + (seq++)));
    fd.opSetOpcode(subOp, CPUI_INT_SUB);
    fd.opSetInput(subOp, bar, 0);
    fd.opSetInput(subOp, fd.newConstant(8, 8), 1);
    Varnode *tC = fd.newUniqueOut(8, subOp);
    fd.opInsertEnd(subOp, block);
    // Consumer that cannot retype tC: a comparison against a constant. (A
    // LOAD consumer would let TypeOpLoad's inslot==-1 reverse edge retype the
    // address as a pointer to the loaded value, masking the null arm.)
    PcodeOp *eqCOp = fd.newOp(2, Address(ram, 0x5000 + (seq++)));
    fd.opSetOpcode(eqCOp, CPUI_INT_EQUAL);
    fd.opSetInput(eqCOp, tC, 0);
    fd.opSetInput(eqCOp, fd.newConstant(8, 0), 1);
    Varnode *vC = fd.newUniqueOut(1, eqCOp);
    fd.opInsertEnd(eqCOp, block);
    // D: INT_ADD pointer arm runs the same propagateAddIn2Out downChain.
    PcodeOp *addOp = fd.newOp(2, Address(ram, 0x5000 + (seq++)));
    fd.opSetOpcode(addOp, CPUI_INT_ADD);
    fd.opSetInput(addOp, bar, 0);
    fd.opSetInput(addOp, fd.newConstant(8, 8), 1);
    Varnode *tD = fd.newUniqueOut(8, addOp);
    fd.opInsertEnd(addOp, block);
    Varnode *vD = makeLoad(tD);
    // E/F: propagateTypeEdge cc:5093 seal check. q is the output of a
    // STOP-flagged PTRSUB from an untyped input (so its local temp stays
    // int8); q2 is the same shape without the stop flag. The INT_EQUAL
    // across-input arm passes bar's plain ProgressData* through unchanged on
    // both sides (typeop.cc:976-981: only the PointerRel form is downgraded):
    // without the seal the sibling input takes the pointer; with the seal the
    // edge is rejected at coreaction.cc:5093.
    Varnode *z = fd.newVarnode(8, Address(reg, 0x140));
    z = fd.setInputVarnode(z);
    Varnode *z2 = fd.newVarnode(8, Address(reg, 0x180));
    z2 = fd.setInputVarnode(z2);
    Varnode *q = [&]() {
      PcodeOp *op = fd.newOp(2, Address(ram, 0x5000 + (seq++)));
      fd.opSetOpcode(op, CPUI_PTRSUB);
      fd.opSetInput(op, z, 0);
      fd.opSetInput(op, fd.newConstant(8, 8), 1);
      Varnode *out = fd.newUniqueOut(8, op);
      fd.opInsertEnd(op, block);
      op->setStopTypePropagation();
      return out;
    }();
    Varnode *q2 = [&]() {
      PcodeOp *op = fd.newOp(2, Address(ram, 0x5000 + (seq++)));
      fd.opSetOpcode(op, CPUI_PTRSUB);
      fd.opSetInput(op, z2, 0);
      fd.opSetInput(op, fd.newConstant(8, 8), 1);
      Varnode *out = fd.newUniqueOut(8, op);
      fd.opInsertEnd(op, block);
      return out;
    }();
    PcodeOp *eqOp = fd.newOp(2, Address(ram, 0x5000 + (seq++)));
    fd.opSetOpcode(eqOp, CPUI_INT_EQUAL);
    fd.opSetInput(eqOp, bar, 0);
    fd.opSetInput(eqOp, q, 1);
    Varnode *cEQ = fd.newUniqueOut(1, eqOp);
    fd.opInsertEnd(eqOp, block);
    PcodeOp *eq2Op = fd.newOp(2, Address(ram, 0x5000 + (seq++)));
    fd.opSetOpcode(eq2Op, CPUI_INT_EQUAL);
    fd.opSetInput(eq2Op, bar, 0);
    fd.opSetInput(eq2Op, q2, 1);
    Varnode *cEQ2 = fd.newUniqueOut(1, eq2Op);
    fd.opInsertEnd(eq2Op, block);

    cells.push_back(Cell{"tA", tA});
    cells.push_back(Cell{"tB", tB});
    cells.push_back(Cell{"tC", tC});
    cells.push_back(Cell{"tD", tD});
    cells.push_back(Cell{"vA", vA});
    cells.push_back(Cell{"vB", vB});
    cells.push_back(Cell{"vC", vC});
    cells.push_back(Cell{"vD", vD});
    cells.push_back(Cell{"q", q});
    cells.push_back(Cell{"q2", q2});
    cells.push_back(Cell{"cEQ", cEQ});
    cells.push_back(Cell{"cEQ2", cEQ2});

    cout << "pre|" << snapshot(cells) << '\n';

    fd.startTypeRecovery();
    ActionInferTypes action("typerecovery");
    action.reset(fd);
    vector<Datatype *> firstTypes;
    for (size_t i = 0; i < cells.size(); ++i)
      firstTypes.push_back(cells[i].vn->getType());
    for (int4 pass = 1; pass <= 8; ++pass) {
      action.apply(fd);
      if (pass == 1) {
        for (size_t i = 0; i < cells.size(); ++i)
          firstTypes[i] = cells[i].vn->getType();
        cout << "pass1|" << snapshot(cells) << '\n';
      }
      else if (pass == 2)
        cout << "pass2|" << snapshot(cells) << '\n';
      else if (pass == 8) {
        bool stable = true;
        for (size_t i = 0; i < cells.size(); ++i)
          stable &= (cells[i].vn->getType() == firstTypes[i]);
        cout << "pass8|" << snapshot(cells)
             << "|types_stable_since_pass1=" << (stable ? 1 : 0) << '\n';
      }
    }
  }
  catch (const LowlevelError &error) {
    cout << "exception|what=" << error.explain << '\n';
    return 1;
  }
  catch (const exception &error) {
    cout << "exception|what=" << error.what() << '\n';
    return 1;
  }
  return 0;
}

File: tests/oracle/stop_ptrsub_wire_1204.rs
/* STOP-PTRSUB-WIRE-0001: ActionInferTypes STOP seal + PTRSUB downChain wiring
 * (Rugra comparand). Mirrors tests/oracle/stop_ptrsub_wire_1204.cc case for
 * case; see that file for the graph design commentary.
 */
use std::sync::{Arc, RwLock};

use rugra::action::Action;
use rugra::address::Address;
use rugra::arch::Architecture;
use rugra::coreaction::ActionInferTypes;
use rugra::funcdata::Funcdata;
use rugra::op::op_addl_flags;
use rugra::opcodes::OpCode;
use rugra::space::AddressSpace;
use rugra::type_system::datatype::{Datatype, TypeBase, TypeField, TypeMetatype, TypeStruct};
use rugra::type_system::typefactory::{CoreTypeFlavor, TypeFactory};
use rugra::varnode::Varnode;

type VarnodeRef = Arc<RwLock<Varnode>>;

fn meta_token(meta: TypeMetatype) -> String {
    match meta {
        TypeMetatype::Void => "void",
        TypeMetatype::Pointer => "ptr",
        TypeMetatype::Array => "array",
        TypeMetatype::Struct => "struct",
        TypeMetatype::Spacebase => "spacebase",
        TypeMetatype::Bool => "bool",
        TypeMetatype::Int => "int",
        TypeMetatype::Uint => "uint",
        _ => "other",
    }
    .to_string()
}

/// Normalized projection shared with the C++ side (no temporary ids, no
/// factory-specific base-type names).
fn type_proj(ct: Option<&Arc<Datatype>>) -> String {
    let Some(ct) = ct else { return "null".to_string() };
    if let Datatype::Pointer(p) = ct.as_ref() {
        let mut res = format!("ptr{}->{}", p.base.size, type_proj(Some(&p.ptr_to)));
        if let Some(rel) = &p.base.pointer_rel {
            res += &format!("|rel={}:{}", rel.parent.get_name(), rel.offset);
        }
        return res;
    }
    format!("{}{}", meta_token(ct.get_metatype()), ct.get_size())
}

struct Cell {
    label: &'static str,
    vn: VarnodeRef,
}

fn snapshot(cells: &[Cell]) -> String {
    cells
        .iter()
        .map(|cell| {
            let vn = cell.vn.read().unwrap();
            format!(
                "{}:{}/stop={}",
                cell.label,
                type_proj(vn.v_type.as_ref()),
                u8::from(vn.stops_up_propagation())
            )
        })
        .collect::<Vec<_>>()
        .join(",")
}

fn main() {
    println!(
        "schema=1|fixture=STOP-PTRSUB-WIRE-0001|oracle=e40ed13014025f82488b1f8f7bca566894ac376b"
    );

    let type_factory = Arc::new(RwLock::new(TypeFactory::new_flavor(
        8,
        CoreTypeFlavor::Standalone,
    )));
    let (progress, progress_pointer) = {
        let mut factory = type_factory.write().unwrap();
        // Mirror the C++ fixture's setupSizes+cacheCoreTypes core-type path:
        // findAdd's alignment gate requires the default map.
        factory.set_default_alignment_map();
        let long_type = factory
            .get_base(8, TypeMetatype::Int)
            .expect("canonical int8");
        let progress = Arc::new(Datatype::Struct(TypeStruct {
            base: TypeBase::new("ProgressData".into(), 32, TypeMetatype::Struct),
            fields: vec![
                TypeField {
                    name: "total".into(),
                    offset: 0,
                    type_ptr: long_type.clone(),
                },
                TypeField {
                    name: "prev".into(),
                    offset: 8,
                    type_ptr: long_type.clone(),
                },
                TypeField {
                    name: "point".into(),
                    offset: 16,
                    type_ptr: long_type,
                },
            ],
        }));
        let pointer = factory.get_type_pointer(8, progress.clone(), 1);
        (progress, pointer)
    };
    let _ = progress;

    let mut architecture = Architecture::new();
    architecture.archid = "x86:LE:64:default:gcc".to_string();
    architecture.set_types(type_factory.clone());
    let mut fd = Funcdata::new("stop_ptrsub_wire", Address::new(0x5000), 0x40);
    fd.vbank.set_type_factory(type_factory.clone());
    fd.set_arch(Arc::new(architecture));
    let block = fd.create_new_block();

    let bar = fd.vbank.create_with_space(8, AddressSpace::Register, 0x100);
    let bar = fd.set_input_varnode(bar);
    bar.write()
        .unwrap()
        .update_type_lock(progress_pointer.clone(), true, false);

    let space_const = fd.new_constant(4, AddressSpace::Ram.space_id() as u64);
    let mut cells: Vec<Cell> = Vec::new();
    let mut seq: u64 = 0;

    let make_ptrsub = |fd: &mut Funcdata,
                       seq: &mut u64,
                       block: &Arc<RwLock<dyn rugra::block::FlowBlock + Send + Sync>>,
                       in0: &VarnodeRef,
                       off: u64,
                       stop: bool|
     -> VarnodeRef {
        let op = fd.new_op(2, Address::new(0x5000 + *seq));
        *seq += 1;
        fd.op_set_opcode(&op, OpCode::CPUI_PTRSUB);
        fd.op_set_input(&op, in0.clone(), 0);
        let cst = fd.new_constant(8, off);
        fd.op_set_input(&op, cst, 1);
        let out = fd.new_unique_out(8, &op);
        fd.op_insert_end(&op, block);
        if stop {
            // RulePtrArith/RuleStructOffset0's bare `addlflags |= 0x40` idiom
            // (ruleaction.cc:6517); op.rs has no dedicated setter.
            op.0.write().unwrap().addlflags |= op_addl_flags::STOP_TYPE_PROPAGATION;
        }
        out
    };
    let make_load = |fd: &mut Funcdata,
                     seq: &mut u64,
                     block: &Arc<RwLock<dyn rugra::block::FlowBlock + Send + Sync>>,
                     space: &VarnodeRef,
                     addr: &VarnodeRef|
     -> VarnodeRef {
        let op = fd.new_op(2, Address::new(0x5000 + *seq));
        *seq += 1;
        fd.op_set_opcode(&op, OpCode::CPUI_LOAD);
        fd.op_set_input(&op, space.clone(), 0);
        fd.op_set_input(&op, addr.clone(), 1);
        let out = fd.new_unique_out(8, &op);
        fd.op_insert_end(&op, block);
        out
    };
    let make_input = |fd: &mut Funcdata, offset: u64| -> VarnodeRef {
        let vn = fd.vbank.create_with_space(8, AddressSpace::Register, offset);
        fd.set_input_varnode(vn)
    };
    let make_equal = |fd: &mut Funcdata,
                      seq: &mut u64,
                      block: &Arc<RwLock<dyn rugra::block::FlowBlock + Send + Sync>>,
                      in0: &VarnodeRef,
                      in1: &VarnodeRef|
     -> VarnodeRef {
        let op = fd.new_op(2, Address::new(0x5000 + *seq));
        *seq += 1;
        fd.op_set_opcode(&op, OpCode::CPUI_INT_EQUAL);
        fd.op_set_input(&op, in0.clone(), 0);
        fd.op_set_input(&op, in1.clone(), 1);
        let out = fd.new_unique_out(1, &op);
        fd.op_insert_end(&op, block);
        out
    };

    // A: sealed PTRSUB + LOAD.
    let t_a = make_ptrsub(&mut fd, &mut seq, &block, &bar, 8, true);
    let v_a = make_load(&mut fd, &mut seq, &block, &space_const, &t_a);
    // B: same PTRSUB without the stop flag.
    let t_b = make_ptrsub(&mut fd, &mut seq, &block, &bar, 8, false);
    let v_b = make_load(&mut fd, &mut seq, &block, &space_const, &t_b);
    // C: INT_SUB null arm (typeop.cc:317-321 base returns null).
    let sub_op = fd.new_op(2, Address::new(0x5000 + seq));
    seq += 1;
    fd.op_set_opcode(&sub_op, OpCode::CPUI_INT_SUB);
    fd.op_set_input(&sub_op, bar.clone(), 0);
    let cst8 = fd.new_constant(8, 8);
    fd.op_set_input(&sub_op, cst8, 1);
    let t_c = fd.new_unique_out(8, &sub_op);
    fd.op_insert_end(&sub_op, &block);
    // Consumer that cannot retype tC (a LOAD's reverse edge would mask the
    // null arm — see the C++ side commentary).
    let zero = fd.new_constant(8, 0);
    let v_c = make_equal(&mut fd, &mut seq, &block, &t_c, &zero);
    // D: INT_ADD pointer arm (same propagateAddIn2Out downChain).
    let add_op = fd.new_op(2, Address::new(0x5000 + seq));
    seq += 1;
    fd.op_set_opcode(&add_op, OpCode::CPUI_INT_ADD);
    fd.op_set_input(&add_op, bar.clone(), 0);
    let cst8b = fd.new_constant(8, 8);
    fd.op_set_input(&add_op, cst8b, 1);
    let t_d = fd.new_unique_out(8, &add_op);
    fd.op_insert_end(&add_op, &block);
    let v_d = make_load(&mut fd, &mut seq, &block, &space_const, &t_d);
    // E/F: cc:5093 seal check through the INT_EQUAL across-input arm.
    let z = make_input(&mut fd, 0x140);
    let z2 = make_input(&mut fd, 0x180);
    let q = make_ptrsub(&mut fd, &mut seq, &block, &z, 8, true);
    let q2 = make_ptrsub(&mut fd, &mut seq, &block, &z2, 8, false);
    let c_eq = make_equal(&mut fd, &mut seq, &block, &bar, &q);
    let c_eq2 = make_equal(&mut fd, &mut seq, &block, &bar, &q2);

    for (label, vn) in [
        ("tA", t_a),
        ("tB", t_b),
        ("tC", t_c),
        ("tD", t_d),
        ("vA", v_a),
        ("vB", v_b),
        ("vC", v_c),
        ("vD", v_d),
        ("q", q),
        ("q2", q2),
        ("cEQ", c_eq),
        ("cEQ2", c_eq2),
    ] {
        cells.push(Cell { label, vn });
    }

    println!("pre|{}", snapshot(&cells));

    fd.start_type_recovery();
    let mut action = ActionInferTypes::new();
    action.reset(&mut fd);
    let mut first_types: Vec<Option<Arc<Datatype>>> = cells
        .iter()
        .map(|cell| cell.vn.read().unwrap().v_type.clone())
        .collect();
    for pass in 1..=8 {
        if let Err(error) = action.apply(&mut fd) {
            println!("exception|what={error}");
            std::process::exit(1);
        }
        match pass {
            1 => {
                first_types = cells
                    .iter()
                    .map(|cell| cell.vn.read().unwrap().v_type.clone())
                    .collect();
                println!("pass1|{}", snapshot(&cells));
            }
            2 => println!("pass2|{}", snapshot(&cells)),
            8 => {
                let stable = cells.iter().zip(first_types.iter()).all(|(cell, first)| {
                    let cur = cell.vn.read().unwrap().v_type.clone();
                    match (&cur, first) {
                        (Some(a), Some(b)) => Arc::ptr_eq(a, b),
                        (None, None) => true,
                        _ => false,
                    }
                });
                println!(
                    "pass8|{}|types_stable_since_pass1={}",
                    snapshot(&cells),
                    u8::from(stable)
                );
            }
            _ => {}
        }
    }
}
